[HLSL][clang] Add elementwise builtin for atan2 (p3)

tex3d opened the pull request:
This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

- Add HLSL frontend for atan2
- Add clang Builtin, map to new llvm.atan2
- SemaChecking restrict to floating point and 2 args
- SemaHLSL restrict to float or half.
- Add to clang ReleaseNotes.rst and LanguageExtensions.rst

Part 3 for Implement the atan2 HLSL Function #70096.

tex3d committed:
[HLSL][clang] Add elementwise builtin for atan2 (p3)

This change is part of this proposal: https://discourse.llvm.org/t/rfc-all-the-math-intrinsics/78294

- Add HLSL frontend for atan2
- Add clang Builtin, map to new llvm.atan2
- SemaChecking restrict to floating point and 2 args
- SemaHLSL restrict to float or half.
- Add to clang ReleaseNotes.rst and LanguageExtensions.rst

spall reviewed (approved):
Code LGTM; didn't look at the tests too closely.

farzonl commented on clang/test/SemaHLSL/BuiltIns/half-float-only-errors2.hlsl:
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_atan2
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_pow
```
remove the https://github.com/llvm/llvm-project/blob/main/clang/test/SemaHLSL/BuiltIns/pow-errors.hlsl file since you are adding this.

farzonl commented on clang/test/SemaHLSL/BuiltIns/half-float-only-errors2.hlsl:
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_atan2
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_pow
+
```
add 
```suggestion
// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_fmod
```
delete the  https://github.com/llvm/llvm-project/blob/main/clang/test/SemaHLSL/BuiltIns/fmod-errors.hlsl file

tex3d replied:
I can do this, but `fmod-errors.hlsl` tests a few extra types (`bool`, `int2`, scalar `double`).  You suggest adding the scalar double case below, but shouldn't we do that for `half-float-only-errors.hlsl` as well?  Then, should we add `bool` and `int` tests to both as well?

Furthermore, I noticed that with the `bool` argument (from fmod test), the error is:
> 1st argument must be a vector, integer or floating point type (was 'bool')

`integer` should not be suggested here as a valid type, so I think there is another fix to be made there.

farzonl replied:
the only HLSL specific checks are to exclude double the rest of those tests have nothing to do with SemaHLSL

farzonl replied:
we can update `half-float-only-errors.hlsl` in a different NFC PR.

tex3d replied:
Sorry, I updated that one here.  Is that ok?

farzonl commented on clang/test/SemaHLSL/BuiltIns/half-float-only-errors2.hlsl:
```diff
@@ -0,0 +1,7 @@
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_atan2
+// RUN: %clang_cc1 -finclude-default-header -triple dxil-pc-shadermodel6.6-library %s -fnative-half-type -emit-llvm-only -disable-llvm-passes -verify -DTEST_FUNC=__builtin_elementwise_pow
+
+double2 test_double_builtin(double2 p0, double2 p1) {
```
```suggestion
double test_double_builtin(double p0, double p1) {
  return TEST_FUNC(p0, p1);
  // expected-error@-1 {{passing 'double' to parameter of incompatible type 'float'}}
}

double2 test_vec_double_builtin(double2 p0, double2 p1) {
```

tex3d committed:
Update tests based on feedback

Remove fmod-errors.hlsl and pow-errors.hlsl, covered in half-float-only-errors2.hlsl.
Add scalar case to half-float-only-errors*.hlsl as well.

zeroomega commented:
We are seeing `Clang :: Sema/aarch64-sve-vector-trig-ops.c` test failure after this change got landed.

```
Exit Code: 1

Command Output (stderr):
--
RUN: at line 1: /b/s/w/ir/x/w/llvm_build/bin/clang -cc1 -internal-isystem /b/s/w/ir/x/w/llvm_build/lib/clang/20/include -nostdsysteminc -triple aarch64 -target-feature +sve    -disable-O0-optnone -o - -fsyntax-only /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Sema/aarch64-sve-vector-trig-ops.c -verify
+ /b/s/w/ir/x/w/llvm_build/bin/clang -cc1 -internal-isystem /b/s/w/ir/x/w/llvm_build/lib/clang/20/include -nostdsysteminc -triple aarch64 -target-feature +sve -disable-O0-optnone -o - -fsyntax-only /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Sema/aarch64-sve-vector-trig-ops.c -verify
error: 'expected-error' diagnostics expected but not seen: 
  File /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Sema/aarch64-sve-vector-trig-ops.c Line 27 (directive at /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Sema/aarch64-sve-vector-trig-ops.c:28): 1st argument must be a vector, integer or floating point type
error: 'expected-error' diagnostics seen but not expected: 
  File /b/s/w/ir/x/w/llvm-llvm-project/clang/test/Sema/aarch64-sve-vector-trig-ops.c Line 27: 1st argument must be a floating point type (was 'svfloat32_t' (aka '__SVFloat32_t'))
2 errors generated.

--
```

There are multiple llvm test failures going on so it is a bit difficult to isolate every one of them. But this patch touched the failing test file. Could you take a look please?
